### ISSUE TITLE
inappropriate narrowing during `match` statement

### ISSUE THREAD

**Issue by bukzor** [microsoft/pyright#8219]
**Describe the bug**
`demo.py`
```python
from typing import assert_type
def demo(args: tuple[str, ...]):
    match args:
        case first, "second":
            print(first, "...")
        case (first,):
            assert_type(first, str)
            print(first)
        case other:
            raise AssertionError(other)
```

Expected behavior:

```console
$ pyright demo.py
0 errors, 0 warnings, 0 informations
```

Actual behavior:

```console
$ pyright demo.py
demo.py
  demo.py:6:14 - error: Pattern will never be matched for subject type "tuple[str, str]" (reportUnnecessaryComparison)
  demo.py:7:25 - error: "assert_type" mismatch: expected "str" but received "Never" (reportAssertTypeFailure)
2 errors, 0 warnings, 0 informations

$ pyright --version
pyright 1.1.368
```

It looks to me like the first `case` is causing a type-narrowing (from `tuple[str, ...]` to `tuple[str, str]`) that persists during the second `case`, which would be incorrect.

**Comment by bukzor** [comment]
Notably substituting a variable-length tuple with `list` causes the program to typecheck successfully:

```python
from typing import assert_type
def demo(args: list[str]):
    match args:
        case (first, "second"):
            print(first)
        case (first,):
            assert_type(first, str)
            print(first)
        case other:
            raise AssertionError(other)


demo(["x"])
demo(["yy", "second"])
```

**Comment by erictraut** [comment]
This is included in pyright 1.1.369.

**Comment by bukzor** [comment]
Thank you!
